### ISSUE TITLE
[feature] programmatic page creation

### ISSUE THREAD

**Issue by m4rrc0** [sw-yx/ssg#19]
So I've spent a little bit of time trying to see if I could make ssg work like Gatsby for page creation (with the `createPage` node API hook).

My approach is currently
1. Fetch data from my CMS in `createIndex` and cache it
2. Use `this.ssgData` in the `preload` function of my `[slug].svelte` component

The main problem with this approach is that the route matching is dependent on file structure. My `[slug].svelte` component can only match routes at level 1. If I want sub-pages I need to create a component at `[one]/[two].svelte`. The problem is I would like to be able to create an indefinite number of nested routes on my external CMS without having to worry about how many sub-folders I created in the source code.

Another problem (arguably) that is obvious with the way Sapper create files on `export` is it only creates files for routes that can be crawled from the index page. We cannot create 'hidden' pages or landing pages for example.

Is it something you are interested to handle with ssg @sw-yx ?
For now, ssg seems to be focused on having your data along with your source code. If we want to enable programmable page creation it seems we are going to diverge from Sapper's core principles in major ways.

**Comment by m4rrc0** [comment]
About the second problem, I just discovered the (new?) `--entry` [option](https://sapper.svelte.dev/docs/#How_it_works) that I think allows us to specify entry points for our app to be crawled from.
If we implement programmatic routes creation, we could use this option to specify all created routes...

**Comment by swyxio** [comment]
hey! yes. my intention is to depart from Sapper sooner or later. Quite frankly i have been dragging my feet because it will be a lot of work. but yeah a programmatic `createPages` api makes sense.

if you have a way to implement it in this project and need it to get your stuff done, be my guest. i can't promise i'll get to the ssg rewrite soon since it is just a hobby project

**Comment by m4rrc0** [comment]
Sure! I understand.
To be honest I wouldn't know where to begin to do that with Sapper.
I have a feeling that a new svelte ssg implementation might be around the corner with your recent experiments and the guys active on the svelvet projet ([for reference](https://github.com/jakedeichert/svelvet/issues/31)).
If I get to the point where I really need this, I think I will just hack around it with nested folders like I mention above for now.

**Comment by swyxio** [comment]
yes, exactly. too early to do this now. it would actually NOT be that hard, if you read the Sapper source code [you can figure out where to insert it](https://github.com/sveltejs/sapper/blob/910d28e3419409b5498933e8c13d37cc03b652a0/src/api/export.ts#L147-L164) but it takes time and energy haha. lets just make sure we have this api when we do the rewrite
